### ISSUE TITLE
Since new assumptions `refine` is horribly slow (2000x slower)

### ISSUE THREAD

**Issue by bjodah** [sympy/sympy#9714]
I noticed a dramatic slow-down of `refine` for some work I was doing.
So I wrote a benchmark and ran it using the new benchmark suite of SymPy:
https://github.com/sympy/sympy_benchmarks/pull/12

Here are the results:
http://hera.physchem.kth.se/~sympy_asv/refine/#refine.TimeRefine01.time_refine

It looks as if #2508 is the source of this performance regression.
Hopefully the new assumptions have plenty of room for optimizations...

@asmeurer @certik @debugger22

**Comment by debugger22** [comment]
#2508 introduced a fallback mechanism which is being used in `ask` and that's certainly the reason for the slowdown. We do hope to improve its performance.

**Comment by asmeurer** [comment]
Yep, 2508 is slower, but more powerful. It was merged partially in hopes that someone would figure out how to make it faster.

**Comment by certik** [comment]
@asmeurer does this PR use an SAT solver that's slow? If so, we can use picosat.

**Comment by asmeurer** [comment]
@bjodah I'd love to hear more about how you are using the new assumptions. One of my assumptions when merging #2508 was that very few things were using the new assumptions, since they work for so few things compared to the old assumptions (`refine` in particular works for almost nothing).

**Comment by asmeurer** [comment]
@certik we are still using the SymPy SAT solver, but remember that pycosat didn't really speed things up that much https://github.com/sympy/sympy/pull/2508#issuecomment-40906245

**Comment by certik** [comment]
@asmeurer thanks, I recall the conversation now. Is it slow because Python is slow, or do you think we can get it faster even in Python?

**Comment by asmeurer** [comment]
I think we can definitely get faster in Python. One issue is that we have some inefficient algorithms in the logic module (like `to_cnf`, which is used by `satisfiable`). See for instance https://github.com/cjdrake/pyeda/blob/master/ipynb/SymPy_Comparison.ipynb. PyEDA is implemented in C, but the issue here is more about using naive algorithms, which in logic often means exponential behavior (read: slow even for small inputs). I chatted with @cjdrake (the author of PyEDA) a bit about this at SciPy. 

That isn't to say that SymEngine couldn't help. I would like to see how SymEngine logic classes would perform (see https://github.com/sympy/sympy/pull/2508#issuecomment-40906332 and https://github.com/sympy/symengine/issues/154).

**Comment by debugger22** [comment]
I created a [PR](https://github.com/sympy/sympy/pull/9813) using @certik's branch with some required changes which uses pycosat for satisfiability but there's no improvement in the speed.

**Comment by bjodah** [comment]
I got around finishing a notebook I've been writing on for a while, so
@asmeurer here is an example of what I use refine for:

http://nbviewer.ipython.org/0b794b1b216534193c02

If you try to run that notebook using SymPy master instead of 0.7.6 it essentially takes hours instead of seconds (which in my humble opinion should be considered a breakage of backward compatibility).

**Comment by moorepants** [comment]
Note that this has gotten even slower with SymPy 1.0:

http://www.moorepants.info/misc/sympy-asv/#refine.TimeRefine01.time_refine
